src/lib/provable/gadgets/sha2.ts

```diff
@@ -0,0 +1,621 @@
+      0x67332667ffc00b31n,
+      0x8eb44a8768581511n,
+      0xdb0c2e0d64f98fa7n,
+      0x47b5481dbefa4fa4n,
+    ],
+    // SHA2-512 §5.3.5
+    512: [
+      0x6a09e667f3bcc908n,
+      0xbb67ae8584caa73bn,
+      0x3c6ef372fe94f82bn,
+      0xa54ff53a5f1d36f1n,
+      0x510e527fade682d1n,
+      0x9b05688c2b3e6c1fn,
+      0x1f83d9abfb41bd6bn,
+      0x5be0cd19137e2179n,
+    ],
+  },
+
+  K: {
+    224: K224_256,
+    256: K224_256,
+    384: K384_512,
+    512: K384_512,
+  },
+};
+
+const SHA2 = {
+  /**
+   * Implementation of [NIST SHA-2](https://csrc.nist.gov/pubs/fips/180-4/upd1/final)
+   * hash Function. Supports output lengths of 224, 256, 384, or 512 bits.
+   *
+   * Applies the SHA-2 hash function to a list of big-endian byte-sized {@link Field}
+   * elements, flexible to handle varying output lengths (224, 256, 384, 512 bits) as specified.
+   *
+   * The function accepts {@link Bytes} as the input message, which is a type that
+   * represents a static-length list of byte-sized field elements (range-checked
+   * using {@link Gadgets.rangeCheck8}).
+   * Alternatively, you can pass plain `number[]` of `Uint8Array` to perform a hash
+   * outside provable code.
+   *
+   * Produces an output of {@link Bytes} that conforms to the chosen bit length.
+   * Both input and output bytes are big-endian.
+   *
+   * @param len - Desired output length in bits. Valid options: 224, 256, 384, 512.
+   * @param message - Big-endian {@link Bytes} representing the message to hash.
+   *
+   * ```ts
+   * let preimage = Bytes.fromString("hello world");
+   * let digest224 = SHA2.hash(224, preimage);
+   * let digest256 = SHA2.hash(256, preimage);
+   * let digest384 = SHA2.hash(384, preimage);
+   * let digest512 = SHA2.hash(512, preimage);
+   * ```
+   *
+   */
+  hash<T extends Length>(length: T, data: FlexibleBytes): Bytes {
+    // Infer the type T based on the value of `length` (conditional type)
+    type Type = T extends 224 | 256 ? UInt32 : UInt64;
+
+    // preprocessing §6.2
+    // padding the message $5.1.1 into blocks that are a multiple of 512
+    let messageBlocks = padding<Type>(length, data);
+
+    let H = SHA2.initialState<Type>(length);
+
+    const N = messageBlocks.length;
+
+    for (let i = 0; i < N; i++) {
+      const W = messageSchedule(length, messageBlocks[i]);
+      H = compression(length, H, W);
+    }
+
+    // the working variables H[i] are 32 | 64 bit, however we want to decompose
+    // them into bytes to be more compatible
+    let digest = Bytes.from(H.map((x) => x.toBytesBE()).flat());
+
+    // Take the first `length` bits of the digest. This has no effect in
+    // SHA256 and SHA512, because 8 words of 32 | 64 bits needs no truncation.
+    // Truncation is required for SHA224 and SHA384 though.
+    digest.bytes = digest.bytes.slice(0, length / 8);
+
+    return digest;
+  },
+  compression,
+  messageSchedule,
+  padding,
+  initialState<T extends UInt32 | UInt64>(length: Length): T[] {
```

Comment:
Unrelated to the PR: is there a convention in o1js to order the methods? As an engineer not working often in TS, I was surprised to see the method `initialState` defined after a place it is used (`hash` above).

Reply:
Not really a convention, TS doesn't care - I personally try to order methods by factors such as importance or level of abstraction